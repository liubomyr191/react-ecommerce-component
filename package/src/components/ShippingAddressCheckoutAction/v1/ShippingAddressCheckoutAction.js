import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { addTypographyStyles, CustomPropTypes } from "../../../utils";

const Title = styled.h3`
  ${addTypographyStyles("ShippingAddressCheckoutActionTitle", "subheadingTextBold")}
`;

const Address = styled.address`
  ${addTypographyStyles("ShippingAddressCheckoutActionAddress", "bodyText")}
`;

class ShippingAddressCheckoutAction extends Component {
  static propTypes = {
    /**
     * If you've set up a components context using
     * [@reactioncommerce/components-context](https://github.com/reactioncommerce/components-context)
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      /**
       * Pass either the Reaction AddressForm component or your own component that
       * accepts compatible props.
       */
      AddressForm: CustomPropTypes.component.isRequired
    }).isRequired,
    /**
     * Checkout data needed for form
     */
    fulfillmentGroup: PropTypes.shape({
      data: PropTypes.shape({
        shippingAddress: CustomPropTypes.address
      })
    }),
    /**
     * Is the shipping address being saved
     */
    isSaving: PropTypes.bool,
    /**
     * Label of workflow step
     */
    label: PropTypes.string.isRequired,
    /**
     * When action is ready for save call this prop method to
     * enable the save button with in the `CheckoutActions`
     */
    onReadyForSaveChange: PropTypes.func,
    /**
     * When an action form passes validation and submits
     * the value will be passed to this callback
     * this function needs to return a Promise
     * if being used with reacto-form
     */
    onSubmit: PropTypes.func,
    /**
     * Checkout process step number
     */
    stepNumber: PropTypes.number.isRequired
  };

  static defaultProps = {
    isSaving: false,
    onReadyForSaveChange() {}
  };

  state = {};

  _addressForm = null;

  submit = () => {
    this._addressForm.submit();
  };

  handleSubmit = async (value) => {
    const { onSubmit } = this.props;
    await onSubmit(value);
  };

  handleChange = (values) => {
    const { onReadyForSaveChange } = this.props;
    const isFilled = Object.keys(values).every((key) => (key === "address2" ? true : values[key] !== null));
    onReadyForSaveChange(isFilled);
  };

  render() {
    const { components: { AddressForm }, fulfillmentGroup, isSaving, label, stepNumber } = this.props;
    const shippingAddress = fulfillmentGroup ? fulfillmentGroup.data.shippingAddress : null;
    return (
      <Fragment>
        <Title>
          {stepNumber}. {label}
        </Title>
        <AddressForm
          ref={(formEl) => {
            this._addressForm = formEl;
          }}
          isSaving={isSaving}
          onChange={this.handleChange}
          onSubmit={this.handleSubmit}
          value={shippingAddress}
        />
      </Fragment>
    );
  }
}

const WrappedShippingAddressCheckoutAction = withComponents(ShippingAddressCheckoutAction);

// eslint-disable-next-line
WrappedShippingAddressCheckoutAction.renderComplete = ({ fulfillmentGroup: { data: { shippingAddress } } }) => (
  <Address>
    {shippingAddress.fullName}
    <br />
    {shippingAddress.address1}
    <br />
    {shippingAddress.address2 !== null && shippingAddress.address2 !== "" ? (
      <span>
        {shippingAddress.address2} <br />
      </span>
    ) : null}
    {shippingAddress.city}, {shippingAddress.region} {shippingAddress.postal} <br />
    {shippingAddress.country}
  </Address>
);

export default WrappedShippingAddressCheckoutAction;
