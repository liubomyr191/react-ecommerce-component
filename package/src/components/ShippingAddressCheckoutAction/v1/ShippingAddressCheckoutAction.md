### Overview
The `ShippingAddressCheckoutAction` serves 4 purposes:
  * To normalize data input/editing between the `CheckoutActions` component and the `AddressForm` components.
  * To provide `CheckoutActions` with a `renderComplete` method that renders the captured checkout data.
  * To provide `CheckoutActions` a way to know the action is ready for save,
  * To Provide `CheckoutActions` a way to capture the `AddressForm` values.

For a full implementation example see the [CheckoutActions](/#!/CheckoutActions)

### Usage

```jsx
const fullfillment = {
  data: {
    shippingAddress: null
  }
};

<ShippingAddressCheckoutAction fulfillmentGroup={fullfillment} label="Shipping Address" stepNumber={1} />
```

#### With Address
```jsx
const fullfillment = {
  data: {
    shippingAddress: {
      address1: "7742 Hwy 23",
      address2: "",
      country: "US",
      city: "Belle Chasse",
      fullName: "Salvos Seafood",
      postal: "70037",
      region: "LA",
      phone: "(504) 393-7303"
    }
  }
};

<ShippingAddressCheckoutAction fulfillmentGroup={fullfillment} label="Shipping Address" stepNumber={1} />
```

#### Without Address
```jsx
const fullfillment = {
  data: {
    shippingAddress: null
  }
};


<ShippingAddressCheckoutAction fulfillmentGroup={fullfillment} label="Shipping Address" stepNumber={1} />
```

#### Ready For Save
Open the browser terminal and fill out this form to see the `onReadyForSaveChange` callback fire.
```jsx
const fullfillment = {
  data: {
    shippingAddress: null
  }
};


const isReady = (ready) => console.log("is action ready for save?", ready);

<ShippingAddressCheckoutAction fulfillmentGroup={fullfillment} onReadyForSaveChange={isReady} label="Shipping Address" stepNumber={1} />
```

#### Render Complete Method
```jsx
const capturedData = {
  fulfillmentGroup: {
    data: {
     shippingAddress: {
        address1: "7742 Hwy 23",
        address2: "",
        country: "US",
        city: "Belle Chasse",
        fullName: "Salvos Seafood",
        postal: "70037",
        region: "LA",
        phone: "(504) 393-7303"
      }
    }
  }
};
const action = ShippingAddressCheckoutAction;

action.renderComplete(capturedData);

```

### Implementation Examples
A naive exmaple to demonstrate a basic shipping address checkout action UX flow. 
**Basic set shipping address**
```jsx
initState = {
  isReady: false,
  fulfillmentGroup: {
    data: {
      shippingAddress: null
    }
  }
};

const action = ShippingAddressCheckoutAction;
let _form = null;

const handleSubmit = (value) => new Promise((resolve, reject) => {
  setState({ isProcessing: true });
  setTimeout(async () => {
    setState({
      isProcessing: false,
      fulfillmentGroup: {
        data: {
          shippingAddress: value
        }
      }
    });
    resolve(value);
  }, 2000);
});

const props = {
  fulfillmentGroup: state.fulfillmentGroup,
  label: "Shipping Address",
  stepNumber: 1,
  onSubmit: handleSubmit,
  onReadyForSaveChange() { setState({ isReady: true }); }
};

<div>
  {state.fulfillmentGroup && state.fulfillmentGroup.data.shippingAddress 
    ? action.renderComplete(state)
    : (<div>
      <ShippingAddressCheckoutAction {...props} ref={(el) => {_form = el}} /> 
      <Button onClick={() => {_form.submit()}} isDisabled={!state.isReady} isWaiting={state.isProcessing}>Submit</Button>
    </div>)}
</div>
```

**With simple address validation**
A naive example to demonstrate a validation shipping address checkout action UX flow. 
```jsx
initState = {
  addressValidationResults: null,
  isReady: false,
  isProcessing: false,
  fulfillmentGroup: {
    data: {
      shippingAddress: null
    }
  }
};

const action = ShippingAddressCheckoutAction;
let _form = null;

const salvos = {
  address1: "7742 Hwy 23",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  fullName: "Salvos Seafood",
  postal: "70037",
  region: "LA",
  phone: "(504) 393-7303"
};

const handleSubmit = (value) => new Promise((resolve, reject) => {
  setState({ isProcessing: true });
  setTimeout(async () => {
    setState({
      isProcessing: false,
      fulfillmentGroup: {
        data: {
          shippingAddress: value
        }
      }
    });
    resolve(value);
  }, 2000);
});

const handleValidation = (value) => new Promise((resolve, reject) => {
  setState({ isProcessing: true });
  setTimeout(async () => {
    setState({
      addressValidationResults: {
        submittedAddress: value,
        suggestedAddresses: [salvos],
        validationErrors: [{
          details: "Sorry but we believe you meant to enter the address of Salvos Seafood",
          type: "warning"
        }]
      },
      isProcessing: false
    });
    resolve(value);
  }, 2000);
});

const props = {
  addressValidationResults: state.addressValidationResults,
  fulfillmentGroup: state.fulfillmentGroup,
  label: "Shipping Address",
  stepNumber: 1,
  onSubmit: handleSubmit,
  onReadyForSaveChange() { setState({ isReady: true }); },
  validation: handleValidation
};

<div>
  {state.fulfillmentGroup && state.fulfillmentGroup.data.shippingAddress 
    ? action.renderComplete(state)
    : (<div>
      <ShippingAddressCheckoutAction {...props} ref={(el) => {_form = el}} /> 
      <Button onClick={() => {_form.submit()}} isDisabled={!state.isReady} isWaiting={state.isProcessing}>Submit</Button>
    </div>)}
</div>
```

### Theme

See [Theming Components](./#!/Theming%20Components).

#### Typography

- The label text uses `subheadingTextBold` style with `rui_components.ShippingAddressCheckoutActionTitle` override
- The address preview text uses `bodyText` style with `rui_components.ShippingAddressCheckoutActionAddress` override
