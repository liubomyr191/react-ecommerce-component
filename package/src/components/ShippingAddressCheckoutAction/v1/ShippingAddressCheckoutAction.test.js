import React from "react";
import renderer from "react-test-renderer";
import ShippingAddressCheckoutAction from "./ShippingAddressCheckoutAction";

test("basic snapshot", () => {
  const component = renderer.create(<ShippingAddressCheckoutAction />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
