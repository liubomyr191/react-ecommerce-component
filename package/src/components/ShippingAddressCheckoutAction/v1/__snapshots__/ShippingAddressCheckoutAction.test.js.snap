// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot with address 1`] = `
Array [
  .c0 {
  font-size: 18px;
  font-weight: 700;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.25;
  -webkit-letter-spacing: 0.4px;
  -moz-letter-spacing: 0.4px;
  -ms-letter-spacing: 0.4px;
  letter-spacing: 0.4px;
}

<h3
    className="c0"
  >
    . 
  </h3>,
  "AddressForm(undefined)",
]
`;

exports[`basic snapshot with empty address 1`] = `
Array [
  .c0 {
  font-size: 18px;
  font-weight: 700;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.25;
  -webkit-letter-spacing: 0.4px;
  -moz-letter-spacing: 0.4px;
  -ms-letter-spacing: 0.4px;
  letter-spacing: 0.4px;
}

<h3
    className="c0"
  >
    . 
  </h3>,
  "AddressForm(undefined)",
]
`;
